core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -415,105 +439,107 @@ public HashDistributionTaskSource(
             this.catalogRequirement = requireNonNull(catalogRequirement, "catalogRequirement is null");
             this.targetPartitionSourceSizeInBytes = requireNonNull(targetPartitionSourceSize, "targetPartitionSourceSize is null").toBytes();
             this.targetPartitionSplitWeight = targetPartitionSplitWeight;
+            this.executor = requireNonNull(executor, "executor is null");
         }
 
         @Override
-        public List<TaskDescriptor> getMoreTasks()
+        public synchronized ListenableFuture<List<TaskDescriptor>> getMoreTasks()
         {
             if (finished || closed) {
-                return ImmutableList.of();
+                return immediateFuture(ImmutableList.of());
             }
+            checkState(loadedSplitsFuture == null || loadedSplitsFuture.isDone(), "getMoreTasks called again before the previous batch of splits was ready");
```

Comment:
Is the `loadedSplitsFuture.isDone()` check necessary?

Reply:
I've removed it now since we're loading everything in one go here